### PR TITLE
Fix silu accuracy

### PR DESCRIPTION
**Pull request by nmauriceTT**
### Ticket
#34861

### Problem description
Accuracy of silu is causing issue with Qwen3 32B accuracy.

### What's changed

This PR fixes silu, which becomes accurate within 1 ULP of bfloat16 and within 5 ULP on float32.

Define silu(x) as `x * sigmoid(x)`. sigmoid(x) is accurate within 1 ULP for bfloat16 and 4 for float32 -> this makes silo(x) accurate as well.

#### Accuracy

silu-bfloat16

silu-float32

Note: There's an outlier with a high ULP error near -87. This is probably a problem with subnormals (to check and possibly fix). Either way, new implementation is more accurate in every way than current one. 

#### Performance

Wormhole_b0:

Type | Branch | Cycles/Datum | Cycles/Tile
-- | -- | -- | --
bfloat16 | main | 1.533844 | 1571
bfloat16 | new | 1.92 | 1971
float32 | main | 1.65 | 1689
float32 | new | 3.38 | 3462

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes [OK](https://github.com/tenstorrent/tt-metal/actions/runs/20402083103)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) [OK](https://github.com/tenstorrent/tt-metal/actions/runs/20402081125)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable) [OK](https://github.com/tenstorrent/tt-metal/actions/runs/20402077511)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable) [OK, failures in main](https://github.com/tenstorrent/tt-metal/actions/runs/20402088516)
- [x] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) [OK, failures in main](https://github.com/tenstorrent/tt-metal/actions/runs/20402079924)
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) job passes
- [ ] New/Existing tests provide coverage for changes
- [x] (Single card) Frequent ttnn model tests : [OK, failures in main/unrelated](https://github.com/tenstorrent/tt-metal/actions/runs/20402407001)

### CODE REVIEW

File: tt_metal/include/compute_kernel_api.h [hunk 1]
```diff
@@ -443,7 +443,7 @@ ALWI void expm1_tile_init() { MATH((llk_math_eltwise_unary_sfpu_expm1_init<true>
 // Function SILU (same as Swish)
 // use activation Silu[x] = x*Sigmoid[x]
 // Ref: https://pytorch.org/docs/stable/generated/torch.nn.SiLU.html?highlight=silu#torch.nn.SiLU
-ALWI void silu_tile(uint32_t idst) { MATH((llk_math_eltwise_unary_sfpu_silu<APPROX>(idst))); }
+ALWI void silu_tile(uint32_t idst) { MATH((llk_math_eltwise_unary_sfpu_silu<APPROX, DST_ACCUM_MODE>(idst))); }
```
Comment (ncvetkovicTT) [inline tt_metal/include/compute_kernel_api.h:457]: Can we use the opportunity to beautify our docstrings here?

```
// clang-format off
/**
 * Performs SILU (same as Swish) operation on each element of a tile
 * in DST register at index tile_index. Uses the following implementation:
 * Silu[x] = x*Sigmoid[x]
 *
 * Ref: https://pytorch.org/docs/stable/generated/torch.nn.SiLU.html?highlight=silu#torch.nn.SiLU
 *
 * Return value: None
 *
 * | Argument        | Description                                                                | Type     | Valid Range                                           | Required |
 * |-----------------|----------------------------------------------------------------------------|----------|-------------------------------------------------------|----------|
 * | idst            | The index of the tile in DST register buffer to perform the computation on | uint32_t | Must be less than the size of the DST register buffer | True     |
 */
// clang-format on
```

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sigmoid.h [hunk 2]
```diff
@@ -64,6 +63,10 @@ inline void calculate_sigmoid() {
             sfpi::vFloat val = sfpi::dst_reg[0];
             sfpi::vFloat result = _sfpu_sigmoid_<is_fp32_dest_acc_en>(val);
 
+            if constexpr (!is_fp32_dest_acc_en) {
```
Comment (fvranicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sigmoid.h:66]: Could you explain what do we gain by moving this check and op out of the `_sfpu_sigmoid_`? This is not affecting accuracy, right? It's more of a style change.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_silu.h [hunk 3]
```diff
@@ -4,36 +4,32 @@
...
+        }
+
+        sfpi::dst_reg[0] = result;
         sfpi::dst_reg++;
     }
 }
 
+template <bool APPROXIMATION_MODE>
+inline void silu_init() {
+    sigmoid_init<false>();
```
Comment (fvranicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_silu.h]: Passing 
```suggestion
    sigmoid_init<APPROXIMATION_MODE>();
```
Or if we don't want to pass approximation mode, simplify `silu_init()` into a simple function without a template parameter. It may be confusing for the user if they are allowed to set different approximation mode, which is then not respected at the first level, inside the function itself.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sigmoid.h (hunk 2 above)
Comment (nmauriceTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_sigmoid.h:66]: For `silu(x) = x * sigmoid(x)`, if we left rounding inside `_sfpu_sigmoid_` then we would have had 2 options:
- `x * _sfpu_sigmoid<false>(x)` -> might lose accuracy from rounding before multiplying. 
- `x * _sfpu_sigmoid<true>(x)` -> gives fp32 accuracy, which may not be needed if dealing with bfloat16 inputs, but would have a performance overhead. 

This change does not affect accuracy (not one else is calling `_sfpu_sigmoid_`, and if they were, then they would probably perform this rounding themselves).

Looking back into it, I think this explicit rounding should only be done right before storing value into `dst_reg` (i.e. in `calculate_sigmoid()` or similar functions).

File: tt_metal/include/compute_kernel_api.h (hunk 1 above)
Comment (nmauriceTT) [reply tt_metal/include/compute_kernel_api.h:457]: Done. Docstring should now be prettier.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_silu.h (hunk 3 above)
Comment (nmauriceTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_silu.h]: I've removed the `APPROXIMATION_MODE` template parameter from both `silu_init()` and `calculate_silu()` (it's not used in either).

File: tt_metal/include/compute_kernel_api.h (hunk 1 above)
Comment (ncvetkovicTT) [reply tt_metal/include/compute_kernel_api.h:457]: They pretty, me likes ^^

**Review by fvranicTT — APPROVED** [review]
